### ISSUE TITLE
GetObject returns invalid metadata from disk cache when using GCS with s3 gateway

### ISSUE THREAD

**Issue by retinio** [minio/minio#8942]
After updated the metadata "get_object" method returns old data if the object is already in disk cache.
I'm running minio as `s3` gateway for GCS with disk caching enabled. 
**docker-compose.yml**
```
version: '3.7'
minio:
    image: minio/minio
    container_name: minio
    volumes:
      - data-1:/data1
    ports:
      - "9000:9000"
    environment:
      MINIO_ACCESS_KEY: acces-key
      MINIO_SECRET_KEY: secret-key
      MINIO_CACHE: "on"
      MINIO_CACHE_DRIVES: "/data1/cache"
      MINIO_CACHE_EXPIRY: 14
      MINIO_CACHE_QUOTA: 80 
    restart: always      
    command: gateway s3 https://storage.googleapis.com

volumes:
  data-1:
```
I'm using AWS SDK for .NET with MinIO Server.
```
const string bucketName = "test-container-1";
const string objectName = "586191.webp";

var request = new CopyObjectRequest
                {
                    SourceBucket = bucketName,
                    SourceKey = objectName,
                    DestinationBucket = bucketName,
                    DestinationKey = objectName,
                    MetadataDirective = S3MetadataDirective.REPLACE
                };
request.Metadata.Add("watermark", "true");
var result = await s3Client.CopyObjectAsync(request );

var requestMeta = new GetObjectMetadataRequest { BucketName = bucketName, Key = objectName,  };
//The response contains correct metadata
var responseMetaBefore = await s3Client.GetObjectMetadataAsync(requestMeta);

var requestGetObject = new GetObjectRequest {BucketName = bucketName, Key = objectName};
//The response contains incorrect metadata 
var firstResponseGetObject = await s3Client.GetObjectAsync(requestGetObject);

//The response contains correct metadata 
var secondResponseGetObject = await s3Client.GetObjectAsync(requestGetObject);
```
## Current Behavior

Metadata in GCS is updated successfully. Metadata in disk cache is updated after the method get_object is run but old data is returned.

## Steps to Reproduce (for bugs)

With minio running as `s3` gateway for GCS with disk caching enabled, using the copy_object method with "x-amz-metadata-directive: REPLACE" to update metadata. Using the get_object method for get old metadata.

## Your Environment

* Version used (`minio version`): RELEASE.2020-01-25T02-50-51Z
* Server type and version: Docker version 19.03.5

**Comment by retinio** [comment]
@kannappanr @poornas 
 This bug is not fixed. The bug has been reproduced. (building from master branch)

**Comment by harshavardhana** [comment]
How did you build from master branch? @retinio

**Comment by retinio** [comment]
In GoLand jetbrains
![image](https://user-images.githubusercontent.com/36562344/74538924-d69a5e00-4f45-11ea-9ed8-6eaefae259a2.png)

**Comment by harshavardhana** [comment]
Your build process is wrong @retinio you are not getting the latest changes here - just build it as shown below

```
GO111MODULE=on go get github.com/minio/minio@master
```

To get the latest

**Comment by retinio** [comment]
@harshavardhana 
Sorry, I don’t understand how your recommendations differ my build process
- git clone https://github.com/minio/minio.git
- building in IDE (GoLand)

I also cloned the repository and build the image from the Dockerfile. The bug is reproduced.

**Comment by harshavardhana** [comment]
> I also cloned the repository and build the image from the Dockerfile. The bug is reproduced.

The `go build` paths in your Goland config are wrong there is no such thing as `go build disk-cache.go` - To properly build the binary you should either use the command I posted or use Dockerfile.dev

@retinio because I can't reproduce the bug, so it means your build process is wrong.

**Comment by retinio** [comment]
@harshavardhana 
My real `go build` paths is `go build github.com/minio/minio` not `go build disk-cache.go`
Working directory contains the latest changes.
![image](https://user-images.githubusercontent.com/36562344/74639449-9d950000-5176-11ea-9098-887cd8bcf780.png)



I'am using [aws-cli](https://github.com/aws/aws-cli/) 
Steps to reproduce:
1. Get metadata before change 
`aws s3api head-object --endpoint-url http://192.168.190.211:9000/  --bucket test-container-1 --key 586191.webp`
{
    "AcceptRanges": "bytes",
    "LastModified": "Tue, 04 Feb 2020 13:09:28 GMT",
    "ContentLength": 5232,
    "ETag": "\"1705ad1b442eb5599e18d5ac9dc55016\"",
    "CacheControl": "private, max-age=0",
    "ContentType": "application/octet-stream",
    "Expires": "Mon, 17 Feb 2020 08:48:02 GMT",
    "Metadata": {
        **"Watermark": "false"**
    }
}

2. Update metadata
`aws s3api copy-object --endpoint-url http://192.168.190.211:9000/ --copy-source test-container-1/586191.webp --key 586191.webp --bucket test-container-1 --metadata-directive REPLACE --metadata Watermark=true`
{
    "CopyObjectResult": {
        "ETag": "\"1705ad1b442eb5599e18d5ac9dc55016\"",
        "LastModified": "2020-02-04T13:09:28.000Z"
    }
}

3. Check changes
`aws s3api head-object --endpoint-url http://192.168.190.211:9000/  --bucket test-container-1 --key 586191.webp`
{
    "AcceptRanges": "bytes",
    "LastModified": "Tue, 04 Feb 2020 13:09:28 GMT",
    "ContentLength": 5232,
    "ETag": "\"1705ad1b442eb5599e18d5ac9dc55016\"",
    "CacheControl": "private, max-age=0",
    "ContentType": "application/octet-stream",
    "Expires": "Mon, 17 Feb 2020 08:53:05 GMT",
    "Metadata": {
        **"Watermark": "true"**
    }
}

4. get-object returns invalid metadata from disk cache
`aws s3api get-object --endpoint-url http://192.168.190.211:9000/  --bucket test-container-1 --key 586191.webp 586191.webp`
{
    "AcceptRanges": "bytes",
    "LastModified": "Mon, 17 Feb 2020 08:44:33 GMT",
    "ContentLength": 5232,
    "ETag": "\"1705ad1b442eb5599e18d5ac9dc55016\"",
    "CacheControl": "private, max-age=0",
    "ContentType": "application/octet-stream",
    "Expires": "Mon, 17 Feb 2020 08:44:33 GMT",
    "Metadata": {
        **"Watermark": "false"**
    }
}

**Comment by harshavardhana** [comment]
Here is how I tested

```
~  aws s3api --profile minio --endpoint-url http://localhost:9000  put-object --bucket testbucket1 --key 1.txt --body /etc/hosts --metadata watermark=true
{
    "ETag": "\"fbde0cdbe51349946d72750a19932e79\""
}
```

```
~ aws s3api --profile minio --endpoint-url http://localhost:9000 head-object --bucket testbucket1 --key 1.txt
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Mon, 17 Feb 2020 11:25:34 GMT", 
    "ContentLength": 282, 
    "ETag": "\"fbde0cdbe51349946d72750a19932e79\"", 
    "ServerSideEncryption": "AES256", 
    "Metadata": {
        "Watermark": "true"
    }
}
```

```
~ aws s3api --profile minio --endpoint-url http://localhost:9000 copy-object --copy-source testbucket1/1.txt --bucket testbucket1 --key 1.txt --metadata-directive REPLACE --metadata watermark=false
{
    "CopyObjectResult": {
        "LastModified": "2020-02-17T11:27:17.000Z", 
        "ETag": "\"20000f008dee695d1855ed8c4d72d7ca-1\""
    }
}
```

```
~ aws s3api --profile minio --endpoint-url http://localhost:9000 get-object --bucket testbucket1 --key 1.txt 1
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Mon, 17 Feb 2020 11:27:17 GMT", 
    "ContentLength": 282, 
    "ETag": "\"20000f008dee695d1855ed8c4d72d7ca-1\"", 
    "ServerSideEncryption": "AES256", 
    "Metadata": {
        "Watermark": "false"
    }
}
```

Gateway started as 

```
~ MINIO_CACHE_QUOTA=99 MINIO_CACHE_DRIVES=/tmp/1 AWS_PROFILE=minio minio gateway s3 --address ":9000" http://localhost:9001
Cache Capacity: 5.7 GiB Free, 44 GiB Total
Endpoint:  http://192.168.225.64:9000  http://172.17.0.1:9000  http://172.18.0.1:9000  http://172.19.0.1:9000  http://127.0.0.1:9000            
AccessKey: minio 
SecretKey: minio123 
...
```

So clearly I don't see the same issue as you do, so it must be that your build process is wrong. AFAIK , so to clear confusion avoid using `GoLand`  just fire up a terminal and perform 

```
~ GO111MODULE=on go get github.com/minio/minio@master
```

**Comment by retinio** [comment]
@harshavardhana I am not a golang developer. I will wait for the next release with fix #8991 and I try to reproduce this bug. Thank you for your time.
